### PR TITLE
B-19891 TAC LOA Invalid DB INT

### PR DESCRIPTION
**Pull request by traskowskycaci**
## [Upstream branch](https://github.com/transcom/mymove/pull/13003)
- Also pulled in a commit per Cameron's recommendation from 19926: [adjust line of accounting handler to better handle no loas found](https://github.com/transcom/mymove/commit/8416fb8d6110818f49d419a57b309c143d4f5cad)
- 19926 is not an upstream

## Summary

This work adds in some logic checks that will determine the result of two flags that we want to store in the database. Advana gets an export of this table, so by storing these validity flags in the lines_of_accounting table, Advana will be able to see if any LOA information is missing.

Two columns that we store a validity check for on the `lines_of_accounting` table:
- `valid_hhg_code_for_loa`
- `valid_loa_for_tac`

Criteria for invalid LOA:

- No HHG Program Code
- No match to orders date and LOA eff date range (HHG)
- No match to orders date and current date (NTS)
- Missing any element per attachment in feature (DFAS elements)
- No LOA record found (LOA system ID key), or nothing returned based on existing logic (check logic for application of LOA to an EDI 858 for payment request transmission)

The logic checks that I added were specifically:
- No HHG Program Code
- Missing any element per attachment in feature (DFAS elements). See the screenshot or the attachment on the feature for the DFAS elements.
image

The other checks should be taken care of by the existing LOA and TAC lookup logic.

I also added a migration to move the valid_loa_for_tac column from the transportation_accounting_codes table to the lines_of_accounting table.

To Test:
1. Refresh your DB to pull in the migraiton.
2. Follow the instructions and use examples from #13003 to call the requestLineOfAccounting endpoint. Verify that the response contains:
`validHhgProgramCodeForLoa`
`validLoaForTac`
3. Verify that the `valid_hhg_code_for_loa` and `valid_loa_for_tac` columns in the database are populated with the correct values after calling the endpoint.

### CODE REVIEW

**Review by danieljordan-caci — APPROVED** [review]
Tested and looks good!

Screenshot 2024-06-14 at 3 10 04 PM
Screenshot 2024-06-14 at 3 10 11 PM

File: migrations/app/schema/20240613192433_move_valid_loa_for_tac_col_to_lines_of_accounting_table.up.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- removing valid_loa_for_tac column from transportation_accounting_codes table
+ALTER TABLE transportation_accounting_codes
+DROP COLUMN IF EXISTS valid_loa_for_tac;
+
+-- then adding valid_loa_for_tac column to lines_of_accounting table
+ALTER TABLE lines_of_accounting
+ADD COLUMN IF NOT EXISTS valid_loa_for_tac boolean DEFAULT NULL;
+
+-- Column comments
+COMMENT ON COLUMN lines_of_accounting.valid_loa_for_tac IS 'Result of LOA lookup for a TAC.';
```
Comment (cameroncaci) [inline migrations/app/schema/20240613192433_move_valid_loa_for_tac_col_to_lines_of_accounting_table.up.sql:10]: Could we adjust this comment to point more towards the service object? Something like "Result of LOA service object validation that occurs whenever new LOAs are fetched by the client"

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go [hunk 2]
```diff
@@ -60,3 +68,144 @@ func sortTransportationAccountingCodesAndLinesOfAccounting(tacs []models.Transpo
...
+		}
+		if loa.LoaFmsTrnsactnID == nil { // Transaction Type
+			missingLoaFields = append(missingLoaFields, "loa.LoaFmsTrnsactnID")
+		}
+		if loa.LoaDscTx == nil { // Foreign Military Sales (FMS) Line Item Number
+			missingLoaFields = append(missingLoaFields, "loa.LoaDscTx")
+		}
+		if loa.LoaUic == nil { // Activity Address Code/UIC
+			missingLoaFields = append(missingLoaFields, "loa.LoaUic")
+		}
```
Comment (cameroncaci) [inline pkg/services/line_of_accounting/line_of_accounting_fetcher.go:171]: Out of convenience for us later down the line, can we add in the DFAS element? Like so

```		// if any LOA DFAS elements are missing, invalid
		var missingLoaFields []string
		if loa.LoaSysID == nil {
			missingLoaFields = append(missingLoaFields, "loa.LoaSysID")
		}
		if loa.LoaDptID == nil { // Department Indicator (A1)
			missingLoaFields = append(missingLoaFields, "loa.LoaDptID")
		}
		if loa.LoaTnsfrDptNm == nil { // Transfer from Department (A2)
			missingLoaFields = append(missingLoaFields, "loa.LoaTnsfrDptNm")
		}
		if loa.LoaBafID == nil { // Basic Symbol Number (A4)
			missingLoaFields = append(missingLoaFields, "loa.LoaBafID")
		}
		if loa.LoaTrsySfxTx == nil { // Subhead/Limit (A5)
			missingLoaFields = append(missingLoaFields, "loa.LoaTrsySfxTx")
		}
		if loa.LoaMajClmNm == nil { // Fund Code/MC (A6)
			missingLoaFields = append(missingLoaFields, "loa.LoaMajClmNm")
		}
		if loa.LoaOpAgncyID == nil { // Operating Agency Code/Fund Admin (B1)
			missingLoaFields = append(missingLoaFields, "loa.LoaOpAgncyID")
		}
		if loa.LoaAlltSnID == nil { // Allotment Serial Number (B2)
			missingLoaFields = append(missingLoaFields, "loa.LoaAlltSnID")
		}
		if loa.LoaPgmElmntID == nil { // Program Element (C1)
			missingLoaFields = append(missingLoaFields, "loa.LoaPgmElmntID")
		}
		if loa.LoaTskBdgtSblnTx == nil { // Project Task or Budget Sub line (C2)
			missingLoaFields = append(missingLoaFields, "loa.LoaTskBdgtSblnTx")
		}
		if loa.LoaDfAgncyAlctnRcpntID == nil { // Defense Agency Allocation Recipient (D1)
			missingLoaFields = append(missingLoaFields, "loa.LoaDfAgncyAlctnRcpntID")
		}
		if loa.LoaJbOrdNm == nil { // Job Order/Work Order Code (D4)
			missingLoaFields = append(missingLoaFields, "loa.LoaJbOrdNm")
		}
		if loa.LoaSbaltmtRcpntID == nil { // Sub-allotment Recipient (D6)
			missingLoaFields = append(missingLoaFields, "loa.LoaSbaltmtRcpntID")
		}
		if loa.LoaWkCntrRcpntNm == nil { // Work Center Recipient (D7)
			missingLoaFields = append(missingLoaFields, "loa.LoaWkCntrRcpntNm")
		}
		if loa.LoaMajRmbsmtSrcID == nil { // Major Reimbursement Source Code (E1)
			missingLoaFields = append(missingLoaFields, "loa.LoaMajRmbsmtSrcID")
		}
		if loa.LoaDtlRmbsmtSrcID == nil { // Detail Reimbursement Source Code (E2)
			missingLoaFields = append(missingLoaFields, "loa.LoaDtlRmbsmtSrcID")
		}
		if loa.LoaCustNm == nil { // Customer Indicator/MPC (E3)
			missingLoaFields = append(missingLoaFields, "loa.LoaCustNm")
		}
		if loa.LoaObjClsID == nil { // Object Class (F1)
			missingLoaFields = append(missingLoaFields, "loa.LoaObjClsID")
		}
		if loa.LoaSrvSrcID == nil { // Government or Public Sector Identifier (F3)
			missingLoaFields = append(missingLoaFields, "loa.LoaSrvSrcID")
		}
		if loa.LoaSpclIntrID == nil { // Special Interest Code or Special Program Cost Code (G2)
			missingLoaFields = append(missingLoaFields, "loa.LoaSpclIntrID")
		}
		if loa.LoaBdgtAcntClsNm == nil { // Abbreviated Department of Defense (DoD) Budget and Accounting (I1)
			missingLoaFields = append(missingLoaFields, "loa.LoaBdgtAcntClsNm")
		}
		if loa.LoaDocID == nil { // (SDN) Document or Record Reference Number (J1)
			missingLoaFields = append(missingLoaFields, "loa.LoaDocID")
		}
		if loa.LoaClsRefID == nil { // (ACRN) Accounting Classification Reference Code (K6)
			missingLoaFields = append(missingLoaFields, "loa.LoaClsRefID")
		}
		if loa.LoaInstlAcntgActID == nil { // Accounting Installation Number (L1)
			missingLoaFields = append(missingLoaFields, "loa.LoaInstlAcntgActID")
		}
		if loa.LoaLclInstlID == nil { // Local Installation Data (M1)
			missingLoaFields = append(missingLoaFields, "loa.LoaLclInstlID")
		}
		if loa.LoaFmsTrnsactnID == nil { // Transaction Type (N1)
			missingLoaFields = append(missingLoaFields, "loa.LoaFmsTrnsactnID")
		}
		if loa.LoaDscTx == nil { // Foreign Military Sales (FMS) Line Item Number (P5)
			missingLoaFields = append(missingLoaFields, "loa.LoaDscTx")
		}
		if loa.LoaUic == nil { // Activity Address Code/UIC (B3)
			missingLoaFields = append(missingLoaFields, "loa.LoaUic")
		}
```

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go [hunk 3]
```diff
@@ -60,3 +68,144 @@ func sortTransportationAccountingCodesAndLinesOfAccounting(tacs []models.Transpo
 
 	return linesOfAccounting
 }
+
+func checkForValidHhgProgramCodeForLoaAndValidLoaForTac(linesOfAccounting []models.LineOfAccounting, appCtx appcontext.AppContext) ([]models.LineOfAccounting, error) {
+	var err error
+	var validHhgProgramCodeForLoa bool
+	var validLoaForTac bool
+	for currLoaIndex, loa := range linesOfAccounting {
```
Comment (cameroncaci) [inline pkg/services/line_of_accounting/line_of_accounting_fetcher.go:76]: DFAS element "A3" is missing. It wasn't present in the crosswalk Dan sent, but A3 is part of the long LOA. See: https://www13.v1host.com/USTRANSCOM38/Default.aspx?Page=Widgets%2fDetails%2fAttachmentDetails&AssetWindow=true&AssetContext=Attachment%3a950468

A3 is the `loaEndFyTx`, and if you're wondering where the "BgFyTx" is, A3 holds the concatenation of both. See snippet below

image

**Review by cameroncaci — CHANGES_REQUESTED** [review]

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go (hunk 3 above)
Comment (traskowskycaci) [reply pkg/services/line_of_accounting/line_of_accounting_fetcher.go:76]: Good catch - thank you! Will get that fix in.

File: migrations/app/schema/20240613192433_move_valid_loa_for_tac_col_to_lines_of_accounting_table.up.sql (hunk 1 above)
Comment (traskowskycaci) [reply migrations/app/schema/20240613192433_move_valid_loa_for_tac_col_to_lines_of_accounting_table.up.sql]: Fixed in commit [9c7a423](https://github.com/transcom/mymove/pull/13070/commits/9c7a4237eb75faa03cf4651cb5d8e7405a43f506)

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go (hunk 2 above)
Comment (traskowskycaci) [reply pkg/services/line_of_accounting/line_of_accounting_fetcher.go]: Fixed in commit [f682e7e](https://github.com/transcom/mymove/pull/13070/commits/f682e7e65c2662348872f5b524924d5085276078)

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go (hunk 3 above)
Comment (traskowskycaci) [reply pkg/services/line_of_accounting/line_of_accounting_fetcher.go:76]: Fixed in commit [9d4b13d](https://github.com/transcom/mymove/pull/13070/commits/9d4b13d524a3e9091c8284b0f22ddb269f391bdb) @cameroncaci

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go [hunk 4]
```diff
@@ -169,6 +169,15 @@ func checkForValidHhgProgramCodeForLoaAndValidLoaForTac(linesOfAccounting []mode
 		if loa.LoaUic == nil { // Activity Address Code/UIC (B3)
 			missingLoaFields = append(missingLoaFields, "loa.LoaUic")
 		}
+		if loa.LoaBgFyTx == nil || loa.LoaEndFyTx == nil { // Ending Fiscal Year Indicator (A3)
+			// LoaEndFyTx = LoaBgFyTx and LoaEndFyTx
+			if loa.LoaBgFyTx == nil {
```
Comment (cameroncaci) [inline pkg/services/line_of_accounting/line_of_accounting_fetcher.go:174]: Sorry, one last thing! Could you set this to be "A3 is a concatenation of both LoaBgFyTx and LoaEndFyTx", just so we don't get confused later down the line cause technically they are separate fields

**Review by cameroncaci — CHANGES_REQUESTED** [review]

File: pkg/services/line_of_accounting/line_of_accounting_fetcher.go (hunk 4 above)
Comment (traskowskycaci) [reply pkg/services/line_of_accounting/line_of_accounting_fetcher.go]: Fixed in commit [update comments and tests](https://github.com/transcom/mymove/pull/13070/commits/a4a37eb13cd419e1f6ec3e73b8c43c0f920dfe34)

Comment (cameroncaci) [reply pkg/services/line_of_accounting/line_of_accounting_fetcher.go]: Fixed in https://github.com/transcom/mymove/pull/13070/commits/a4a37eb13cd419e1f6ec3e73b8c43c0f920dfe34

**Review by cameroncaci — APPROVED** [review]
Good job! Love it 🚀

**Pull request merged**